File: smart-agriculture-suite/src/LearningModule.cpp
#include "LearningModule.h"

#include <math.h>

namespace agri {

namespace {
const int kActionDurations[LearningModule::Count] = {0, 30, 60, 120};
}  // namespace

const char* LearningModule::actionName(Action action) {
    static const char* kNames[] = {"off", "low", "medium", "high"};
    return kNames[static_cast<int>(action)];
}

void LearningModule::begin(const LearningConfig& config) {
    config_ = config;
    loadQTable();
}

void LearningModule::update(const SensorSnapshot& snapshot, bool sampleUpdated, unsigned long nowMs, ActuatorController& actuator) {
    if (!sampleUpdated) {
        return;
    }

    latestSnapshot_ = snapshot;
    const int currentState = discretizeState(snapshot);

    if (hasPendingReward_ && nowMs - lastDecisionMs_ >= config_.decisionIntervalMs) {
        const float reward = calculateReward(pendingSoilBefore_, snapshot.soilHumi, lastAction_);
        lastReward_ = reward;
        totalReward_ += reward;
        averageReward_ = totalEpisodes_ > 0 ? totalReward_ / static_cast<float>(totalEpisodes_) : 0.0f;
        updateQValue(lastState_, lastAction_, reward, currentState);

        LearningRecord& record = history_[historyIndex_ % kHistorySize];
        record.state = lastState_;
        record.action = lastAction_;
        record.reward = reward;
        record.soilBefore = pendingSoilBefore_;
        record.soilAfter = snapshot.soilHumi;
        historyIndex_ = (historyIndex_ + 1) % kHistorySize;
        historyCount_ = min(historyCount_ + 1, kHistorySize);

        hasPendingReward_ = false;
    }

    if (!config_.autoControlEnabled || !actuator.status().autoMode || actuator.isBusy(nowMs)) {
        return;
    }
    if (nowMs - lastDecisionMs_ < config_.decisionIntervalMs) {
        return;
    }

    const Action action = selectAction(currentState);
    if (action != Off) {
        actuator.startTimedRun(ControlSource::Learning, static_cast<unsigned long>(kActionDurations[action]), nowMs);
    }

    pendingSoilBefore_ = snapshot.soilHumi;
    lastState_ = currentState;
    lastAction_ = action;
    hasPendingReward_ = true;
    ++totalEpisodes_;
    lastDecisionMs_ = nowMs;
    // Adaptive epsilon decay: faster decay when average reward is improving
    float decayFactor = config_.epsilonDecay;
    if (totalEpisodes_ > 100 && averageReward_ > 2.0f) {
        decayFactor = config_.epsilonDecay * 0.998f;  // Decay faster when learning well
    }
    config_.epsilon = max(config_.epsilonMin, config_.epsilon * decayFactor);

    if (totalEpisodes_ % 50 == 0) {
        saveQTable();
    }
}

void LearningModule::reset() {
    initQTable();
    totalEpisodes_ = 0;
    totalReward_ = 0.0f;
    averageReward_ = 0.0f;
    lastReward_ = 0.0f;
    lastState_ = 0;
    lastAction_ = Off;
    hasPendingReward_ = false;
    historyCount_ = 0;
    historyIndex_ = 0;
    config_.epsilon = 0.3f;
    prefs_.begin("qlearn", false);
    prefs_.clear();
    prefs_.end();
}

void LearningModule::recordUserFeedback(bool positive) {
    const float feedbackReward = positive ? 5.0f : -5.0f;
    const int state = discretizeState(latestSnapshot_);
    updateQValue(state, lastAction_, feedbackReward, state);
    ++userOverrideCount_;
    userSatisfaction_ = constrain(userSatisfaction_ + (positive ? 2.0f : -3.0f), 0.0f, 100.0f);
}

void LearningModule::writeStatus(JsonDocument& doc) const {
    const int state = discretizeState(latestSnapshot_);

    doc["autoControlEnabled"] = config_.autoControlEnabled;
    doc["airTemp"] = latestSnapshot_.airTemp;
    doc["airHumi"] = latestSnapshot_.airHumi;
    doc["soilHumi"] = latestSnapshot_.soilHumi;
    doc["liquidLevel"] = latestSnapshot_.liquidLevel;
    doc["lightValue"] = latestSnapshot_.lightValue;
    doc["currentState"] = state;
    doc["lastAction"] = static_cast<int>(lastAction_);
    doc["lastActionName"] = actionName(lastAction_);
    doc["lastReward"] = lastReward_;
    doc["averageReward"] = averageReward_;
    doc["totalEpisodes"] = totalEpisodes_;
    doc["epsilon"] = config_.epsilon;
    doc["targetSoil"] = config_.targetSoil;
    doc["userOverrides"] = userOverrideCount_;
    doc["userSatisfaction"] = userSatisfaction_;

    int bestAction = 0;
    for (int a = 1; a < Count; ++a) {
        if (qTable_[state][a] > qTable_[state][bestAction]) {
            bestAction = a;
        }
    }
    doc["recommendedAction"] = actionName(static_cast<Action>(bestAction));

    JsonArray qValues = doc["qValues"].to<JsonArray>();
    for (int a = 0; a < Count; ++a) {
        JsonObject item = qValues.add<JsonObject>();
        item["action"] = actionName(static_cast<Action>(a));
        item["value"] = qTable_[state][a];
    }
}

void LearningModule::writeQTableSummary(JsonDocument& doc) const {
    int nonZero = 0;
    float maxQ = -9999.0f;
    float minQ = 9999.0f;

    for (int s = 0; s < kStateCount; ++s) {
        for (int a = 0; a < Count; ++a) {
            if (fabsf(qTable_[s][a]) > 0.001f) {
                ++nonZero;
                maxQ = max(maxQ, qTable_[s][a]);
                minQ = min(minQ, qTable_[s][a]);
            }
        }
    }

    doc["totalStates"] = kStateCount;
    doc["totalEntries"] = kStateCount * Count;
    doc["nonZeroEntries"] = nonZero;
    doc["coverage"] = (static_cast<float>(nonZero) / static_cast<float>(kStateCount * Count)) * 100.0f;
    doc["maxQ"] = nonZero > 0 ? maxQ : 0.0f;
    doc["minQ"] = nonZero > 0 ? minQ : 0.0f;

    JsonArray history = doc["history"].to<JsonArray>();
    for (int i = 0; i < historyCount_; ++i) {
        const int idx = (historyIndex_ - 1 - i + kHistorySize) % kHistorySize;
        JsonObject item = history.add<JsonObject>();
        item["state"] = history_[idx].state;
        item["action"] = actionName(history_[idx].action);
        item["reward"] = history_[idx].reward;
        item["soilBefore"] = history_[idx].soilBefore;
        item["soilAfter"] = history_[idx].soilAfter;
    }
}

void LearningModule::initQTable() {
    memset(qTable_, 0, sizeof(qTable_));
}

void LearningModule::loadQTable() {
    prefs_.begin("qlearn", true);
    totalEpisodes_ = prefs_.getInt("episodes", 0);
    averageReward_ = prefs_.getFloat("avgReward", 0.0f);
    totalReward_ = averageReward_ * static_cast<float>(totalEpisodes_);
    config_.epsilon = prefs_.getFloat("epsilon", config_.epsilon);

    if (totalEpisodes_ > 0) {
        for (int s = 0; s < kStateCount; ++s) {
            char key[12];
            snprintf(key, sizeof(key), "q%d", s);
            prefs_.getBytes(key, qTable_[s], sizeof(float) * Count);
        }
    } else {
        initQTable();
    }
    prefs_.end();
}

void LearningModule::saveQTable() {
    prefs_.begin("qlearn", false);
    for (int s = 0; s < kStateCount; ++s) {
        char key[12];
        snprintf(key, sizeof(key), "q%d", s);
        prefs_.putBytes(key, qTable_[s], sizeof(float) * Count);
    }
    prefs_.putInt("episodes", totalEpisodes_);
    prefs_.putFloat("avgReward", averageReward_);
    prefs_.putFloat("epsilon", config_.epsilon);
    prefs_.end();
}

int LearningModule::discretizeState(const SensorSnapshot& snapshot) const {
    const int temp = discretizeTemp(snapshot.airTemp);
    const int humi = discretizeHumi(snapshot.airHumi);
    const int soil = discretizeSoil(snapshot.soilHumi);
    const int light = discretizeLight(snapshot.lightValue);
    const int period = getTimePeriod(millis());

    return temp * (kHumiLevels * kSoilLevels * kLightLevels * kTimeLevels) +
           humi * (kSoilLevels * kLightLevels * kTimeLevels) +
           soil * (kLightLevels * kTimeLevels) +
           light * kTimeLevels +
           period;
}

int LearningModule::discretizeTemp(float temp) const {
    if (temp < 10.0f) return 0;
    if (temp < 18.0f) return 1;
    if (temp < 25.0f) return 2;
    if (temp < 33.0f) return 3;
    return 4;
}

int LearningModule::discretizeHumi(float humi) const {
    if (humi < 30.0f) return 0;
    if (humi < 50.0f) return 1;
    if (humi < 70.0f) return 2;
    return 3;
}

int LearningModule::discretizeSoil(float soil) const {
    if (soil < 20.0f) return 0;
    if (soil < 35.0f) return 1;
    if (soil < 50.0f) return 2;
    if (soil < 65.0f) return 3;
    return 4;
}

int LearningModule::discretizeLight(float light) const {
    if (light < 100.0f) return 0;
    if (light < 500.0f) return 1;
    return 2;
}

int LearningModule::getTimePeriod(unsigned long nowMs) const {
    const unsigned long hours = (nowMs / 3600000UL) % 24UL;
    if (hours >= 6UL && hours < 12UL) {
        return 0;
    }
    if (hours >= 12UL && hours < 18UL) {
        return 1;
    }
    return 2;
}

LearningModule::Action LearningModule::selectAction(int state) const {
    if (random(1000) < config_.epsilon * 1000.0f) {
        return static_cast<Action>(random(Count));
    }

    int bestAction = 0;
    float bestValue = qTable_[state][0];
    for (int a = 1; a < Count; ++a) {
        if (qTable_[state][a] > bestValue) {
            bestValue = qTable_[state][a];
            bestAction = a;
        }
    }
    return static_cast<Action>(bestAction);
}

float LearningModule::calculateReward(float soilBefore, float soilAfter, Action action) const {
    float reward = 0.0f;
    const float diffAfter = fabsf(soilAfter - config_.targetSoil);
    const float diffBefore = fabsf(soilBefore - config_.targetSoil);

    // Core reward: proximity to target soil moisture
    if (diffAfter <= config_.soilTolerance) {
        reward += 10.0f;
    } else {
        reward -= diffAfter * 0.3f;
    }

    // Improvement reward: getting closer to target
    if (diffAfter < diffBefore) {
        reward += 3.0f;
    }

    // Efficiency reward: prefer not watering when soil is already adequate
    if (action == Off && soilBefore > config_.targetSoil - config_.soilTolerance) {
        reward += 2.0f;
    }

    // Energy efficiency: prefer lower intensity when possible
    if (action == Low && diffBefore < config_.soilTolerance * 2.0f) {
        reward += 1.5f;  // Reward conservative watering
    }
    if (action == High && diffBefore < config_.soilTolerance) {
        reward -= 1.0f;  // Penalize overwatering with high intensity
    }

    // Safety penalties
    if (soilAfter > 80.0f) {
        reward -= 5.0f;
    }
    if (soilAfter < 20.0f) {
        reward -= 8.0f;
    }

    // Overshooting penalty: went past target
    if (soilBefore < config_.targetSoil && soilAfter > config_.targetSoil + config_.soilTolerance) {
        reward -= 2.0f;
    }

    return reward;
}

void LearningModule::updateQValue(int state, Action action, float reward, int nextState) {
    float maxNextQ = qTable_[nextState][0];
    for (int a = 1; a < Count; ++a) {
        maxNextQ = max(maxNextQ, qTable_[nextState][a]);
    }

    const float currentQ = qTable_[state][action];
    const float newQ = currentQ + config_.alpha * (reward + config_.gamma * maxNextQ - currentQ);
    qTable_[state][action] = newQ;
}

}  // namespace agri

File: smart-agriculture-atlas200dk/src/ai/learning_module.py
"""
智润智慧农业套件 - Atlas 200I DK A2 版
Q-Learning学习模块 - 900状态空间, 4动作, 文件持久化Q表

对应原ESP32项目的 LearningModule.h/LearningModule.cpp
使用JSON文件持久化替代NVS
"""

import math
import time
import json
import random
import logging
from dataclasses import dataclass
from typing import Optional, List

import numpy as np

from config.app_types import SensorSnapshot, ControlSource, IrrigationAction
from config.hardware_config import DATA_DIR

logger = logging.getLogger(__name__)


@dataclass
class LearningConfig:
    auto_control_enabled: bool = True
    decision_interval_sec: float = 300.0  # 5分钟
    epsilon: float = 0.3
    epsilon_min: float = 0.05
    epsilon_decay: float = 0.9995
    alpha: float = 0.1        # 学习率
    gamma: float = 0.95       # 折扣因子
    target_soil: float = 55.0  # 目标土壤湿度
    soil_tolerance: float = 10.0

    # 离散化分档
    temp_bins: int = 5
    humi_bins: int = 4
    soil_bins: int = 5
    light_bins: int = 3
    time_bins: int = 3


ACTION_DURATIONS = [0, 30, 45, 120]  # Off, Low, Moderate, Heavy (秒)
ACTION_NAMES = ["off", "low", "medium", "high"]


class LearningModule:
    """Q-Learning灌溉决策 - 5×4×5×3×3 = 900状态, 4动作"""

    STATE_COUNT = 5 * 4 * 5 * 3 * 3  # 900

    def __init__(self, config: LearningConfig = None, data_dir: str = DATA_DIR):
        self._config = config or LearningConfig()
        self._data_dir = data_dir

        self._q_table = np.zeros((self.STATE_COUNT, 4))
        self._total_episodes = 0
        self._total_reward = 0.0
        self._average_reward = 0.0
        self._last_reward = 0.0
        self._last_state = 0
        self._last_action = IrrigationAction.Off
        self._has_pending_reward = False
        self._pending_soil_before = 0.0
        self._last_decision_time = 0.0
        self._latest_snapshot = SensorSnapshot()

        self._user_override_count = 0
        self._user_satisfaction = 50.0

        # 历史记录 (最近100条)
        self._history: List[dict] = []

    def begin(self):
        self._load_q_table()

    def update(self, snapshot: SensorSnapshot, sample_updated: bool,
               now: float, actuator, prediction_risk: float = 0.0) -> Optional[IrrigationAction]:
        if not sample_updated:
            return None

        self._latest_snapshot = snapshot
        current_state = self._discretize_state(snapshot)

        # 处理上次决策的奖励
        if (self._has_pending_reward and
                now - self._last_decision_time >= self._config.decision_interval_sec):
            reward = self._calculate_reward(
                self._pending_soil_before, snapshot.soil_humi, self._last_action, prediction_risk)
            self._last_reward = reward
            self._total_reward += reward
            self._average_reward = (self._total_reward / self._total_episodes
                                     if self._total_episodes > 0 else 0.0)
            self._update_q_value(self._last_state, self._last_action, reward, current_state)

            self._history.append({
                "state": self._last_state,
                "action": ACTION_NAMES[self._last_action.value],
                "reward": round(reward, 3),
                "soilBefore": round(self._pending_soil_before, 1),
                "soilAfter": round(snapshot.soil_humi, 1),
            })
            if len(self._history) > 100:
                self._history.pop(0)

            self._has_pending_reward = False

        # 检查是否需要做决策
        if (not self._config.auto_control_enabled or
                not actuator._auto_mode or
                actuator.is_busy(now)):
            return None

        if now - self._last_decision_time < self._config.decision_interval_sec:
            return None

        action = self._select_action(current_state)
        if action != IrrigationAction.Off:
            duration = ACTION_DURATIONS[action.value]
            actuator.start_timed_run(ControlSource.TimedRun, duration, now)

        self._pending_soil_before = snapshot.soil_humi
        self._last_state = current_state
        self._last_action = action
        self._has_pending_reward = True
        self._total_episodes += 1
        self._last_decision_time = now

        # 自适应epsilon衰减
        decay_factor = self._config.epsilon_decay
        if self._total_episodes > 100 and self._average_reward > 2.0:
            decay_factor *= 0.998
        self._config.epsilon = max(self._config.epsilon_min,
                                    self._config.epsilon * decay_factor)

        if self._total_episodes % 50 == 0:
            self._save_q_table()

        return action

    def record_user_feedback(self, positive: bool):
        feedback_reward = 5.0 if positive else -5.0
        state = self._discretize_state(self._latest_snapshot)
        self._update_q_value(state, self._last_action, feedback_reward, state)
        self._user_override_count += 1
        self._user_satisfaction = max(0.0, min(100.0,
            self._user_satisfaction + (2.0 if positive else -3.0)))

    def reset(self):
        self._q_table = np.zeros((self.STATE_COUNT, 4))
        self._total_episodes = 0
        self._total_reward = 0.0
        self._average_reward = 0.0
        self._last_reward = 0.0
        self._has_pending_reward = False
        self._history.clear()
        self._config.epsilon = 0.3
        self._save_q_table()

    # ------------------------------------------------------------------
    # 状态离散化
    # ------------------------------------------------------------------
    def _discretize_state(self, snapshot: SensorSnapshot) -> int:
        t = self._discretize_temp(snapshot.air_temp)
        h = self._discretize_humi(snapshot.air_humi)
        s = self._discretize_soil(snapshot.soil_humi)
        l = self._discretize_light(snapshot.light_intensity)
        p = self._get_time_period()

        return (t * (4 * 5 * 3 * 3) +
                h * (5 * 3 * 3) +
                s * (3 * 3) +
                l * 3 + p)

    @staticmethod
    def _discretize_temp(temp: float) -> int:
        if temp < 10: return 0
        if temp < 18: return 1
        if temp < 25: return 2
        if temp < 33: return 3
        return 4

    @staticmethod
    def _discretize_humi(humi: float) -> int:
        if humi < 30: return 0
        if humi < 50: return 1
        if humi < 70: return 2
        return 3

    @staticmethod
    def _discretize_soil(soil: float) -> int:
        if soil < 20: return 0
        if soil < 35: return 1
        if soil < 50: return 2
        if soil < 65: return 3
        return 4

    @staticmethod
    def _discretize_light(light: float) -> int:
        if light < 100: return 0
        if light < 500: return 1
        return 2

    @staticmethod
    def _get_time_period() -> int:
        hour = time.localtime().tm_hour
        if 6 <= hour < 12: return 0
        if 12 <= hour < 18: return 1
        return 2

    # ------------------------------------------------------------------
    # 动作选择与Q值更新
    # ------------------------------------------------------------------
    def _select_action(self, state: int) -> IrrigationAction:
        if random.random() < self._config.epsilon:
            return IrrigationAction(random.randint(0, 3))
        best = int(np.argmax(self._q_table[state]))
        return IrrigationAction(best)

    def _calculate_reward(self, soil_before: float, soil_after: float,
                           action: IrrigationAction, prediction_risk: float = 0.0) -> float:
        cfg = self._config
        diff_after = abs(soil_after - cfg.target_soil)
        diff_before = abs(soil_before - cfg.target_soil)

        reward = 0.0

        if diff_after <= cfg.soil_tolerance:
            reward += 10.0
        else:
            reward -= diff_after * 0.3

        if diff_after < diff_before:
            reward += 3.0

        if action == IrrigationAction.Off and soil_before > cfg.target_soil - cfg.soil_tolerance:
            reward += 2.0

        if action == IrrigationAction.Low and diff_before < cfg.soil_tolerance * 2.0:
            reward += 1.5

        if action == IrrigationAction.Heavy and diff_before < cfg.soil_tolerance:
            reward -= 1.0

        if soil_after > 80.0:
            reward -= 5.0
        if soil_after < 20.0:
            reward -= 8.0

        # 预测风险惩罚: 土壤预测偏低时, 不灌溉会受额外惩罚
        if prediction_risk > 0.3 and action == IrrigationAction.Off:
            reward -= prediction_risk * 2.0

        if soil_before < cfg.target_soil and soil_after > cfg.target_soil + cfg.soil_tolerance:
            reward -= 2.0

        return reward

    def _update_q_value(self, state: int, action: IrrigationAction,
                         reward: float, next_state: int):
        max_next_q = float(np.max(self._q_table[next_state]))
        current_q = self._q_table[state][action.value]
        new_q = current_q + self._config.alpha * (
            reward + self._config.gamma * max_next_q - current_q)
        self._q_table[state][action.value] = new_q

    # ------------------------------------------------------------------
    # 持久化
    # ------------------------------------------------------------------
    def _load_q_table(self):
        import os
        path = os.path.join(self._data_dir, "q_table.json")
        if os.path.exists(path):
            try:
                with open(path, 'r') as f:
                    data = json.load(f)
                self._q_table = np.array(data["q_table"])
                self._total_episodes = data.get("episodes", 0)
                self._average_reward = data.get("avg_reward", 0.0)
                self._config.epsilon = data.get("epsilon", self._config.epsilon)
                self._total_reward = self._average_reward * self._total_episodes
                logger.info(f"Q表已加载: {self._total_episodes}轮")
            except Exception as e:
                logger.warning(f"Q表加载失败: {e}")

    def _save_q_table(self):
        import os
        os.makedirs(self._data_dir, exist_ok=True)
        path = os.path.join(self._data_dir, "q_table.json")
        try:
            with open(path, 'w') as f:
                json.dump({
                    "q_table": self._q_table.tolist(),
                    "episodes": self._total_episodes,
                    "avg_reward": self._average_reward,
                    "epsilon": self._config.epsilon,
                }, f)
            logger.info(f"Q表已保存: {self._total_episodes}轮")
        except Exception as e:
            logger.warning(f"Q表保存失败: {e}")

    def to_dict(self) -> dict:
        state = self._discretize_state(self._latest_snapshot)
        best_action = int(np.argmax(self._q_table[state]))
        return {
            "autoControlEnabled": self._config.auto_control_enabled,
            "currentState": state,
            "lastAction": self._last_action.value,
            "lastActionName": ACTION_NAMES[self._last_action.value],
            "lastReward": round(self._last_reward, 3),
            "averageReward": round(self._average_reward, 3),
            "totalEpisodes": self._total_episodes,
            "epsilon": round(self._config.epsilon, 5),
            "targetSoil": self._config.target_soil,
            "userOverrides": self._user_override_count,
            "userSatisfaction": round(self._user_satisfaction, 1),
            "recommendedAction": ACTION_NAMES[best_action],
            "qValues": [
                {"action": ACTION_NAMES[a], "value": round(float(self._q_table[state][a]), 4)}
                for a in range(4)
            ],
        }
